phpdoc/files/common/settings/Settings.php.txt
<?php
/**
 * Application settings file.  Loads and holds static copy of the default.php specified data.
 *
 * @author          Simon Mitchell <kartano@gmail.com>
 *
 * @namespace       Prometheus2\common\settings
 *
 * @version         1.0.0           2017-08-17 2017-08-17 Prototype
 */

namespace Prometheus2\common\settings;

use Prometheus2\common\logging\logger;

/**
 * Class Settings
 * @package Prometheus2\common\settings
 */
class Settings
{
    public static $config;
    /**
     * @var string $filename File from which configuration settings were loaded.
     */
    protected static $filename = '';

    /**
     * Get the static instance of the config data.
     *
     * @return array An array of settings.
     * @throws \Exception Thrown if there was no useful data in the default config file.
     */
    protected static function getData(): array
    {
        $filename = dirname(__FILE__) . '\\..\\..\\share\\config\\default.json';
        $config = json_decode(file_get_contents($filename), true);
        self::$config = $config;
        self::$config['db']['socket'] = intval(ini_get("mysqli.default_socket"));
        return self::$config;
    }

    /**
     * @param array ...$optionPathComponents The path from which to retrieve a setting.  I.E:
     *                                       Settings::get('db','host');
     *
     * @return mixed The value identified from the list of arguments.
     */
    public static function get(...$optionPathComponents)
    {
        $ret = self::getData();
        foreach ($optionPathComponents as $key) {
            if (!isset($ret[$key])) {
                return null;
            }
            $ret = $ret[$key];
        }
        return $ret;
    }

    /**
     * Set a value in the settings array.  THE VALUE IS NOT PERSISTENT BETWEEN EXECUTIONS!
     *
     * @param mixed $value                   The value to store in specified path.
     * @param array ...$optionPathComponents The path from which to retrieve a setting
     */
    public static function set($value, ...$optionPathComponents)
    {
        $currentOption = self::getData();
        foreach ($optionPathComponents as $key) {
            if (!is_array($currentOption)) {
                $currentOption[$key] = [];
            }
            $currentOption =   &$currentOption[$key];
        }
        $currentOption = $value;
    }

    /**
     * Returns an instance of the specified logger class type for this application.
     *
     * @return \prometheus2\common\logging\logger
     */
    public static function getLogger(): logger
    {
        static $logger = null;
        if ($logger === null) {
            $classType = self::get('log', 'class');
            $logger = new $classType;
        }
        return $logger;
    }
}



phpdoc/files/common/pagerendering/pageoptions.php.txt
<?php
/**
 * Page rendering class options
 *
 * @author          Simon Mitchell <kartano@gmail.com>
 *
 * @namespace       Prometheus2\common\pagerendering
 *
 * @version         1.0.0        2017-08-18 12:39:00 Prototype
 */

/**
 * @property bool   uses_jquery      Flag to determine if header needs to include jquery.
 * @property bool   uses_jqueryui    Flag to determine if header needs to include jquery ui
 * @property bool   display_header   Flag to determine if page <header> element is rendered.
 * @property bool   display_footer   Flag to determine if page <footer> element is rendered.
 * @property string document_type    The HTML Document type.
 * @property string render_body_only If TRUE then only the body of the page is render - the <HEAD> and <BODY> tags do
 *           NOT get used.
 * @property string title            The title of the page.  For the SITE title, use that supplied in SETTINGS.
 * @property string sectionid        The HTML ID.
 * @property string description The Description to be used for the meta tags in web pages.
 * @property bool   requires_logged_in  If TRUE then session authentication will be checked.  THIS WILL *NOT* WORK IF YOU ARE OPERATING WITH CONTENT ONLY RENDERING!
 * @property bool   uses_font_awesome If TRUE then the CDN for the font awesome icon set is imported.  TRUE by default.
 */

namespace Prometheus2\common\pagerendering;

use Prometheus2\common\settings\Settings AS CFG;

/**
 * Class PageOptions
 * @package Prometheus2\common\pagerendering
 */
class PageOptions
{
    /**
     * @var int $sectioncount A static var shared between all page options, to help ensure they are unique.
     */
    protected static $sectioncount = 1;
    /**
     * @var array The array of settings
     */
    protected $data = [];

    /**
     * PageOptions constructor.
     */
    public function __construct()
    {
        $this->uses_jquery = true;
        $this->uses_jqueryui = true;
        $this->display_header = true;
        $this->display_footer = true;

        $this->document_type = 'HTML';

        $this->render_body_only = false;

        $this->title = CFG::get('app', 'name');

        $this->sectionid = 'section' . self::$sectioncount++;

        $this->description = CFG::get('app','description');

        $this->requires_logged_in=false;

        $this->uses_font_awesome=true;
    }

    /**
     * @param string $name The name of the setting.
     *
     * @return mixed The return value. NULL of setting not found.
     */
    public function __get(string $name)
    {
        if (array_key_exists($name, $this->data)) {
            return $this->data[$name];
        } else {
            return null;
        }
    }

    /**
     * Magic method - set.
     *
     * @param string $name  The name of the setting.
     * @param mixed  $value The value for the setting.
     */
    public function __set(string $name, $value)
    {
        $this->data[$name] = $value;
    }
}



common/settings/Settings.php
<?php
/**
 * Application settings file.  Loads and holds static copy of the default.php specified data.
 *
 * @author          Simon Mitchell <kartano@gmail.com>
 *
 * @namespace       Prometheus2\common\settings
 *
 * @version         1.0.0           2017-08-17 2017-08-17 Prototype
 */

namespace Prometheus2\common\settings;

use Prometheus2\common\logging\logger;

/**
 * Class Settings
 * @package Prometheus2\common\settings
 */
class Settings
{
    public static $config;
    /**
     * @var string $filename File from which configuration settings were loaded.
     */
    protected static $filename = '';

    /**
     * Get the static instance of the config data.
     *
     * @return array An array of settings.
     * @throws \Exception Thrown if there was no useful data in the default config file.
     */
    protected static function getData(): array
    {
        $filename = dirname(__FILE__) . '\\..\\..\\share\\config\\default.json';
        $config = json_decode(file_get_contents($filename), true);
        self::$config = $config;
        self::$config['db']['socket'] = intval(ini_get("mysqli.default_socket"));
        return self::$config;
    }

    /**
     * @param array ...$optionPathComponents The path from which to retrieve a setting.  I.E:
     *                                       Settings::get('db','host');
     *
     * @return mixed The value identified from the list of arguments.
     */
    public static function get(...$optionPathComponents)
    {
        $ret = self::getData();
        foreach ($optionPathComponents as $key) {
            if (!isset($ret[$key])) {
                return null;
            }
            $ret = $ret[$key];
        }
        return $ret;
    }

    /**
     * Set a value in the settings array.  THE VALUE IS NOT PERSISTENT BETWEEN EXECUTIONS!
     *
     * @param mixed $value                   The value to store in specified path.
     * @param array ...$optionPathComponents The path from which to retrieve a setting
     */
    public static function set($value, ...$optionPathComponents)
    {
        $currentOption = self::getData();
        foreach ($optionPathComponents as $key) {
            if (!is_array($currentOption)) {
                $currentOption[$key] = [];
            }
            $currentOption =   &$currentOption[$key];
        }
        $currentOption = $value;
    }

    /**
     * Returns an instance of the specified logger class type for this application.
     *
     * @return \prometheus2\common\logging\logger
     */
    public static function getLogger(): logger
    {
        static $logger = null;
        if ($logger === null) {
            $classType = self::get('log', 'class');
            $logger = new $classType;
        }
        return $logger;
    }
}
